File: OneSignalSDK/onesignal/src/main/java/com/onesignal/PushRegistratorHMS.java
package com.onesignal;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.TextUtils;

import com.huawei.agconnect.config.AGConnectServicesConfig;
import com.huawei.hms.aaid.HmsInstanceId;
import com.huawei.hms.common.ApiException;
import com.huawei.hms.push.HmsMessaging;
import com.huawei.hms.support.api.entity.core.CommonCode;

import static com.onesignal.OneSignal.LOG_LEVEL;

class PushRegistratorHMS implements PushRegistrator {

    static final String HMS_CLIENT_APP_ID = "client/app_id";

    private static final int NEW_TOKEN_TIMEOUT_MS = 30_000;

    private static boolean callbackSuccessful;
    private @Nullable static RegisteredHandler registeredHandler;

    static void fireCallback(String token) {
        if (registeredHandler == null)
            return;
        callbackSuccessful = true;
        registeredHandler.complete(token, UserState.PUSH_STATUS_SUBSCRIBED);
    }

    @Override
    public void registerForPush(@NonNull final Context context, final String senderId, @NonNull final RegisteredHandler callback) {
        registeredHandler = callback;
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    getHMSTokenTask(context, callback);
                } catch (ApiException e) {
                    OneSignal.Log(LOG_LEVEL.ERROR, "HMS ApiException getting Huawei push token!", e);

                    int pushStatus;
                    if (e.getStatusCode() == CommonCode.ErrorCode.ARGUMENTS_INVALID)
                        pushStatus = UserState.PUSH_STATUS_HMS_ARGUMENTS_INVALID;
                    else
                        pushStatus = UserState.PUSH_STATUS_HMS_API_EXCEPTION_OTHER;

                    callback.complete(null, pushStatus);
                }
            }
        }, "OS_HMS_GET_TOKEN").start();
    }

    private synchronized void getHMSTokenTask(@NonNull Context context, @NonNull RegisteredHandler callback) throws ApiException {
        // Check required to prevent AGConnectServicesConfig or HmsInstanceId used below
        //   from throwing a ClassNotFoundException
        if (!OSUtils.hasAllHMSLibrariesForPushKit()) {
            callback.complete(null, UserState.PUSH_STATUS_MISSING_HMS_PUSHKIT_LIBRARY);
            return;
        }

        String appId = AGConnectServicesConfig.fromContext(context).getString(HMS_CLIENT_APP_ID);
        HmsInstanceId hmsInstanceId = HmsInstanceId.getInstance(context);

        String pushToken = hmsInstanceId.getToken(appId, HmsMessaging.DEFAULT_TOKEN_SCOPE);

        if (!TextUtils.isEmpty(pushToken)) {
            OneSignal.Log(LOG_LEVEL.INFO, "Device registered for HMS, push token = " + pushToken);
            callback.complete(pushToken, UserState.PUSH_STATUS_SUBSCRIBED);
        } else {
            waitForOnNewPushTokenEvent(callback);
        }
    }

    private static void doTimeOutWait() {
        try {
            Thread.sleep(NEW_TOKEN_TIMEOUT_MS);
        } catch (InterruptedException e) {
        }
    }

    // If EMUI 9.x or older getToken will always return null.
    // We must wait for HmsMessageService.onNewToken to fire instead.
    private void waitForOnNewPushTokenEvent(@NonNull RegisteredHandler callback) {
        doTimeOutWait();
        if (!callbackSuccessful) {
            OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "HmsMessageServiceOneSignal.onNewToken timed out.");
            callback.complete(null, UserState.PUSH_STATUS_HMS_TOKEN_TIMEOUT);
        }
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/notification/internal/registration/PushRegistratorHMS.kt
package com.onesignal.onesignal.notification.internal.registration

import android.content.Context
import android.text.TextUtils
import com.huawei.agconnect.config.AGConnectServicesConfig
import com.huawei.hms.aaid.HmsInstanceId
import com.huawei.hms.common.ApiException
import com.huawei.hms.push.HmsMessaging
import com.huawei.hms.support.api.entity.core.CommonCode
import com.onesignal.onesignal.core.internal.device.IDeviceService
import com.onesignal.onesignal.core.internal.logging.Logging
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout

internal class PushRegistratorHMS(private val _deviceService: IDeviceService) : IPushRegistrator {

    companion object {
        private const val HMS_CLIENT_APP_ID = "client/app_id"
    }

    private var _channel: Channel<String?>? = null

    override suspend fun registerForPush(context: Context) : IPushRegistrator.RegisterResult = coroutineScope {
        var result: IPushRegistrator.RegisterResult? = null

        launch(Dispatchers.Default) {
            result = try {
                getHMSTokenTask(context)
            } catch (e: ApiException) {
                Logging.error("HMS ApiException getting Huawei push token!", e)
                val pushStatus: IPushRegistrator.RegisterStatus =
                    if (e.statusCode == CommonCode.ErrorCode.ARGUMENTS_INVALID)
                        IPushRegistrator.RegisterStatus.PUSH_STATUS_HMS_ARGUMENTS_INVALID
                    else
                        IPushRegistrator.RegisterStatus.PUSH_STATUS_HMS_API_EXCEPTION_OTHER

                IPushRegistrator.RegisterResult(null, pushStatus)
            }
        }

        return@coroutineScope result!!
    }

    @Synchronized
    @Throws(ApiException::class)
    private suspend fun getHMSTokenTask(context: Context) : IPushRegistrator.RegisterResult {
        // Check required to prevent AGConnectServicesConfig or HmsInstanceId used below
        //   from throwing a ClassNotFoundException
        if (!_deviceService.hasAllHMSLibrariesForPushKit()) {
            return IPushRegistrator.RegisterResult(
                null,
                IPushRegistrator.RegisterStatus.PUSH_STATUS_MISSING_HMS_PUSHKIT_LIBRARY
            )
        }

        _channel = Channel()
        val appId = AGConnectServicesConfig.fromContext(context).getString(HMS_CLIENT_APP_ID)
        val hmsInstanceId = HmsInstanceId.getInstance(context)
        var pushToken = hmsInstanceId.getToken(appId, HmsMessaging.DEFAULT_TOKEN_SCOPE)
        if (!TextUtils.isEmpty(pushToken)) {
            Logging.info("Device registered for HMS, push token = $pushToken")
            return IPushRegistrator.RegisterResult(
                pushToken,
                IPushRegistrator.RegisterStatus.PUSH_STATUS_SUBSCRIBED
            )
        } else {
            // If EMUI 9.x or older getToken will always return null.
            // We must wait for HmsMessageService.onNewToken to fire instead.
            // wait up to 30 seconds for someone to call `fireCallback` with the registration id.
            // if it comes before we will continue immediately.
            withTimeout(30000) {
                pushToken = _channel?.receive()
            }

            return if(pushToken != null) {
                Logging.error("ADM registered with ID:$pushToken")
                IPushRegistrator.RegisterResult(
                    pushToken,
                    IPushRegistrator.RegisterStatus.PUSH_STATUS_SUBSCRIBED
                )
            } else {
                Logging.error("HmsMessageServiceOneSignal.onNewToken timed out.")
                IPushRegistrator.RegisterResult(
                    null,
                    IPushRegistrator.RegisterStatus.PUSH_STATUS_HMS_TOKEN_TIMEOUT
                )
            }
        }
    }

    suspend fun fireCallback(id: String?) {
        _channel?.send(id)
    }
}